Added ability to restrict each test to a different room

RitheshShenthar opened the pull request:
The changes made to steeplechase allow tests to run in individual rooms using the same socket by restricting the messaging to specific rooms and exiting those rooms after each test.

RitheshShenthar committed:
Added ability to restrict each test to a different room

nils-ohlmeier commented on webharness/harness.js:
```diff
@@ -76,24 +101,69 @@ function wait_for_message() {
 }
 
 /*
+ * Return a promise for the next available message
```
Could you please update the comment here and on wait_for_test_messsage() so that they don't look like copy and paste any more? ;-)

nils-ohlmeier commented on webharness/harness.js:
```diff
@@ -76,24 +101,69 @@ function wait_for_message() {
 
-function connect_socket() {
+/*
+* Send an object as a message on |socket|.
```
Same as above: please update the comment.

nils-ohlmeier commented on webharness/harness.js:
```diff
@@ -76,24 +101,69 @@ function wait_for_message() {
     server += "/";
   }
-  var room = SpecialPowers.getCharPref("steeplechase.signalling_room");
+  var adminroom = SpecialPowers.getCharPref("steeplechase.signalling_room");
```
This variable over-writes the global var 'adminroom'. Not sure if that is a good idea.

nils-ohlmeier commented on webharness/harness.js:
```diff
@@ -105,7 +175,7 @@ function connect_socket() {
   }).then(function () {
     return new Promise((resolve, reject) => {
-      socket.once("numclients", function(data) {
+      socket.on('numclients', function(data) {
```
Why is this change from once() to on() needed?

nils-ohlmeier commented on webharness/harness.js:
```diff
@@ -145,33 +215,34 @@ function run_next_test() {
+      });
     });
   });
-  //TODO: timeout handling
```
Can we please preserve this comment?

BTW a Promise race would now probably an easy way to implement timeouts here.

nils-ohlmeier commented on webharness/harness.js:
```diff
@@ -184,9 +255,29 @@ addEventListener("error", harness_error);
 // Called by tests via test.js.
 function test_finished() {
+  socket.emit('unsubscribe', testroom);
+  var clients_exited = new Promise((resolve, reject) => {
```
I'm not entirely convinced that waiting for both clients to exit is a good idea. We have already a synchronization point when both clients join the room for the next test and publish their test_loaded message into the admin room. So I don't see much additional value, but extra complexity. (Note: leave the socket.emit() above in place - simply do not worry about if the 'unsubscribe' was successful or not.)

RitheshShenthar committed:
Corrected logical issue with messaging

1) Fixed the messaging system between client and server to include room
information when sending data to server/client.
2) Corrected the logic to start socket.on()s only in the first loop(
when we enter admin room) and not to start multiple instances of
socket.ons in every iteration.

nils-ohlmeier commented:
And can you please rebase your changes as Github tells me it can't automatically merge this any more.

RitheshShenthar committed:
Incorporating Nils comments
